commit 94b63e66206a9ad38872a9136a623ce73cf7c858
Author: Florian Weimer <fweimer@redhat.com>
Date:   Thu Feb 7 09:02:00 2019 +0100

    support: Implement xdlmopen
    
    Put xdlmopen into its own file, to avoid disturbing static linking
    tests (where dlmopen pulls in additional code).

diff --git a/support/Makefile b/support/Makefile
index 5808a42dce87151f..895b83a426369b0c 100644
--- a/support/Makefile
+++ b/support/Makefile
@@ -86,6 +86,7 @@ libsupport-routines = \
   xconnect \
   xcopy_file_range \
   xdlfcn \
+  xdlmopen \
   xdup2 \
   xfclose \
   xfopen \
diff --git a/support/xdlfcn.h b/support/xdlfcn.h
index ab1cbb3cb9bb1cc7..a53fb61b133af5c3 100644
--- a/support/xdlfcn.h
+++ b/support/xdlfcn.h
@@ -25,11 +25,11 @@ __BEGIN_DECLS
 
 /* Each of these terminates process on failure with relevant error message.  */
 void *xdlopen (const char *filename, int flags);
+void *xdlmopen (Lmid_t lmid, const char *filename, int flags);
 void *xdlsym (void *handle, const char *symbol);
 void *xdlvsym (void *handle, const char *symbol, const char *version);
 void xdlclose (void *handle);
 
-
 __END_DECLS
 
 #endif /* SUPPORT_DLOPEN_H */
diff --git a/support/xdlmopen.c b/support/xdlmopen.c
new file mode 100644
index 0000000000000000..9a39ba8801eb1617
--- /dev/null
+++ b/support/xdlmopen.c
@@ -0,0 +1,31 @@
+/* dlmopen with error checking.
+   Copyright (C) 2017-2019 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <http://www.gnu.org/licenses/>.  */
+
+#include <support/check.h>
+#include <support/xdlfcn.h>
+
+void *
+xdlmopen (Lmid_t lmid, const char *filename, int flags)
+{
+  void *dso = dlmopen (lmid, filename, flags);
+
+  if (dso == NULL)
+    FAIL_EXIT1 ("error: dlmopen: %s\n", dlerror ());
+
+  return dso;
+}
